### PR TITLE
Concept of pull-through cache for GitHub calls

### PR DESCRIPTION
**Pull request by jmguzik**
Do not merge: **concept**. It is deployed in my playground namespace, see: `https://gh-pull-through-cache-jguzik-playground.apps.ci.l2s4.p1.openshiftapps.com/`

Cache hits:
```
$ wget --server-response https://gh-pull-through-cache-jguzik-playground.apps.ci.l2s4.p1.openshiftapps.com/golang/dep/releases/download/v0.5.4/dep-linux-amd64
--2022-08-22 15:13:13--  https://gh-pull-through-cache-jguzik-playground.apps.ci.l2s4.p1.openshiftapps.com/golang/dep/releases/download/v0.5.4/dep-linux-amd64
Resolving gh-pull-through-cache-jguzik-playground.apps.ci.l2s4.p1.openshiftapps.com (gh-pull-through-cache-jguzik-playground.apps.ci.l2s4.p1.openshiftapps.com)... 44.198.94.101, 52.72.159.251
Connecting to gh-pull-through-cache-jguzik-playground.apps.ci.l2s4.p1.openshiftapps.com (gh-pull-through-cache-jguzik-playground.apps.ci.l2s4.p1.openshiftapps.com)|44.198.94.101|:443... connected.
HTTP request sent, awaiting response... 
  HTTP/1.1 200 OK
  server: nginx/1.20.1
  date: Mon, 22 Aug 2022 13:13:14 GMT
  content-type: application/octet-stream
  content-length: 11081280
  last-modified: Wed, 08 Dec 2021 06:12:55 GMT
  etag: "0x8D9BA11C66C0428"
  x-ms-request-id: fc9b1a36-801e-0077-4228-b6f7f1000000
  x-ms-version: 2020-04-08
  x-ms-creation-time: Wed, 18 Aug 2021 04:20:21 GMT
  x-ms-lease-status: unlocked
  x-ms-lease-state: available
  x-ms-blob-type: BlockBlob
  content-disposition: attachment; filename=dep-linux-amd64
  x-ms-server-encrypted: true
  fastly-restarts: 1
  age: 134
  via: 1.1 varnish
  x-served-by: cache-iad-kcgs7200072-IAD
  x-cache: HIT
  x-cache-hits: 0
  x-timer: S1661173994.170947,VS0,VE15
  x-cache-status: MISS
  accept-ranges: bytes
  set-cookie: ca352b14fcf2717b0c5d512fc9aa0357=af46a9a8017fa51de6f82b17e4c119fc; path=/; HttpOnly; Secure; SameSite=None
  cache-control: private
Length: 11081280 (11M) [application/octet-stream]
Saving to: ‘dep-linux-amd64’

dep-linux-amd64                                            100%[========================================================================================================================================>]  10.57M  7.11MB/s    in 1.5s    

2022-08-22 15:13:15 (7.11 MB/s) - ‘dep-linux-amd64’ saved [11081280/11081280]

[jguzik@fedora release]$ wget --server-response https://gh-pull-through-cache-jguzik-playground.apps.ci.l2s4.p1.openshiftapps.com/golang/dep/releases/download/v0.5.4/dep-linux-amd64
--2022-08-22 15:13:18--  https://gh-pull-through-cache-jguzik-playground.apps.ci.l2s4.p1.openshiftapps.com/golang/dep/releases/download/v0.5.4/dep-linux-amd64
Resolving gh-pull-through-cache-jguzik-playground.apps.ci.l2s4.p1.openshiftapps.com (gh-pull-through-cache-jguzik-playground.apps.ci.l2s4.p1.openshiftapps.com)... 52.72.159.251, 44.198.94.101
Connecting to gh-pull-through-cache-jguzik-playground.apps.ci.l2s4.p1.openshiftapps.com (gh-pull-through-cache-jguzik-playground.apps.ci.l2s4.p1.openshiftapps.com)|52.72.159.251|:443... connected.
HTTP request sent, awaiting response... 
  HTTP/1.1 200 OK
  server: nginx/1.20.1
  date: Mon, 22 Aug 2022 13:13:18 GMT
  content-type: application/octet-stream
  content-length: 11081280
  last-modified: Wed, 08 Dec 2021 06:12:55 GMT
  etag: "0x8D9BA11C66C0428"
  x-ms-request-id: fc9b1a36-801e-0077-4228-b6f7f1000000
  x-ms-version: 2020-04-08
  x-ms-creation-time: Wed, 18 Aug 2021 04:20:21 GMT
  x-ms-lease-status: unlocked
  x-ms-lease-state: available
  x-ms-blob-type: BlockBlob
  content-disposition: attachment; filename=dep-linux-amd64
  x-ms-server-encrypted: true
  fastly-restarts: 1
  age: 134
  via: 1.1 varnish
  x-served-by: cache-iad-kcgs7200072-IAD
  x-cache: HIT
  x-cache-hits: 0
  x-timer: S1661173994.170947,VS0,VE15
  x-cache-status: HIT
  accept-ranges: bytes
  set-cookie: ca352b14fcf2717b0c5d512fc9aa0357=af46a9a8017fa51de6f82b17e4c119fc; path=/; HttpOnly; Secure; SameSite=None
  cache-control: private
Length: 11081280 (11M) [application/octet-stream]
Saving to: ‘dep-linux-amd64.1’

dep-linux-amd64.1                                          100%[========================================================================================================================================>]  10.57M  7.69MB/s    in 1.4s    

2022-08-22 15:13:20 (7.69 MB/s) - ‘dep-linux-amd64.1’ saved [11081280/11081280]

```

/cc @jupierce @bear-redhat 
/hold

Signed-off-by: Jakub Guzik <jguzik@redhat.com>

### CODE REVIEW

File: clusters/app.ci/gh-pull-through-cache/gh-pull-through-cache.yaml [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+            error_page 301 302 307 = @handle_redirect;
+            resolver 8.8.8.8;
+          }
+        }
+      }
+      events {
+        worker_connections  4096;
+      }
+- apiVersion: route.openshift.io/v1
+  kind: Route
```
Comment (jupierce) [inline clusters/app.ci/gh-pull-through-cache/gh-pull-through-cache.yaml:50]: Might just be here for testing, but ultimately we would not be exposing a Route. Pods to pull through the service URL relative to their build farm. e.g. `http://gh-pull-through-cache.jguzik-playground.svc/<release artifact path>`

Comment (jmguzik) [reply clusters/app.ci/gh-pull-through-cache/gh-pull-through-cache.yaml:50]: Only for the testing. I used my PC to send requests as well.

File: clusters/app.ci/gh-pull-through-cache/gh-pull-through-cache.yaml [hunk 2]
```diff
@@ -0,0 +1,112 @@
+apiVersion: v1
+kind: List
+items:
+- apiVersion: v1
+  kind: ConfigMap
+  metadata:
+    name: gh-pull-through-cache
+    namespace: jguzik-playground
```
Comment (jmguzik) [inline clusters/app.ci/gh-pull-through-cache/gh-pull-through-cache.yaml:8]: @jupierce target namespace would be ci right?

Comment (jupierce) [reply clusters/app.ci/gh-pull-through-cache/gh-pull-through-cache.yaml:8]: Anything should work, but that's where I'd put it.

**Pull request closed without merging**